README
Nageru is a live video mixer, based around the standard M/E workflow.


Features:

 - High performance on modest hardware (720p60 with two input streams
   on my Thinkpad X240[1]); almost all pixel processing is done on the GPU.

 - High output quality; Lanczos3 scaling, subpixel precision everywhere,
   white balance adjustment, mix of 16- and 32-bit floating point
   for intermediate calculations, dithered output, optional 10-bit input
   and output support.

 - Proper sound support: Syncing of multiple unrelated sources through
   high-quality resampling, multichannel mixing with separate effects
   per-bus, cue out for headphones, dynamic range compression,
   three-band graphical EQ (pluss a fixed low-cut), level meters conforming
   to EBU R128, automation via MIDI controllers.

 - Theme engine encapsulating the design demands of each individual
   event; Lua code is responsible for setting up the pixel processing
   pipelines, running transitions etc., so that the visual look is
   consistent between operators.

 - Comprehensive monitoring through Prometheus metrics.

[1] For reference, that is: Core i7 4600U (dualcore 2.10GHz, clocks down
to 800 MHz after 30 seconds due to thermal constraints), Intel HD Graphics
4400 (ie., without the extra L4 cache from Iris Pro), single-channel DDR3 RAM
(so 12.8 GB/sec theoretical memory bandwidth, shared between CPU and GPU).


Nageru is in beta stage. It currently needs:

 - An Intel processor with Intel Quick Sync, or otherwise some hardware
   H.264 encoder exposed through VA-API. Note that you can use VA-API over
   DRM instead of X11, to use a non-Intel GPU for rendering but still use
   Quick Sync (by giving e.g. “--va-display /dev/dri/renderD128”).

 - Two or more Blackmagic USB3 or PCI cards, either HDMI or SDI.
   The PCI cards need Blackmagic's own drivers installed. The USB3 cards
   are driven through the “bmusb” driver, using libusb-1.0. If you want
   zerocopy USB, you need libusb 1.0.21 or newer, as well as a recent
   kernel (4.6.0 or newer). Zerocopy USB helps not only for performance,
   but also for stability. You need at least version 0.7.0.

 - Movit, my GPU-based video filter library (https://movit.sesse.net).
   You will need at least version 1.5.2.

 - Qt 5.5 or newer for the GUI.

 - QCustomPlot for the histogram display in the frame analyzer.

 - libmicrohttpd for the embedded web server.

 - x264 for encoding high-quality video suitable for streaming to end users.

 - ffmpeg for muxing, and for encoding audio. You will need at least
   version 3.1.

 - Working OpenGL; Movit works with almost any modern OpenGL implementation.
   Nageru has been tested with Intel on Mesa (you want 11.2 or newer, due
   to critical stability bugfixes), and with NVIDIA's proprietary drivers.
   AMD's proprietary drivers (fglrx) are known not to work due to driver bugs;
   I am in contact with AMD to try to get this resolved.

 - libzita-resampler, for resampling sound sources so that they are in sync
   between sources, and also for oversampling for the peak meter.

 - Lua, for driving the theme engine.

 - libpci, for printing friendly PCI device names in an error message.


If on Debian stretch or something similar, you can install everything you need
with:

  apt install qtbase5-dev libqt5opengl5-dev qt5-default libqcustomplot-dev \
    pkg-config libmicrohttpd-dev libusb-1.0-0-dev liblua5.2-dev \
    libzita-resampler-dev libva-dev libavcodec-dev libavformat-dev \
    libswscale-dev libavresample-dev libmovit-dev libegl1-mesa-dev \
    libasound2-dev libx264-dev libbmusb-dev protobuf-compiler \
    libprotobuf-dev libpci-dev

Exceptions as of July 2017:

  - You will need Movit from unstable; stretch only has 1.4.0.

  - You will need bmusb from unstable; stretch only has 0.5.4.

The patches/ directory contains a patch that helps zita-resampler performance.
It is meant for upstream, but was not in at the time Nageru was released.
It is taken to be by Steinar H. Gunderson <sesse@google.com> (ie., my ex-work
email), and under the same license as zita-resampler itself.

To start it, just hook up your equipment, type “make” and then “./nageru”.
It is strongly recommended to have the rights to run at real-time priority;
it will make the USB3 threads do so, which will make them a lot more stable.
(A reasonable hack for testing is probably just to run it as root using sudo,
although you might not want to do that in production.) Note also that if you
are running a desktop compositor, it will steal significant amounts of GPU
performance. The same goes for PulseAudio.

Nageru will open a HTTP server at port 9095, where you can extract a live
H264+PCM signal in nut mux (e.g. http://127.0.0.1:9095/stream.nut).
It is probably too high bitrate (~25 Mbit/sec depending on content) to send to
users, but you can easily send it around in your internal network and then
transcode it in e.g. VLC. A copy of the stream (separately muxed) will also
be saved live to local disk.

If you have a fast CPU (typically a quadcore desktop; most laptops will spend
most of their CPU on running Nageru itself), you can use x264 for the outgoing
stream instead of Quick Sync; it is much better quality for the same bitrate,
and also has proper bitrate controls. Simply add --http-x264-video on the
command line. (You may also need to add something like "--x264-preset veryfast",
since the default "medium" preset might be too CPU-intensive, but YMMV.)
The stream saved to disk will still be the Quick Sync-encoded stream, as it is
typically higher bitrate and thus also higher quality. Note that if you add
".metacube" at the end of the URL (e.g. "http://127.0.0.1:9095/stream.ts.metacube"),
you will get a stream suitable for streaming through the Cubemap video reflector
(cubemap.sesse.net). A typical example would be:

  ./nageru --http-x264-video --x264-preset veryfast --x264-tune film \
      --http-mux mp4 --http-audio-codec libfdk_aac --http-audio-bitrate 128

If you are comfortable with using all your remaining CPU power on the machine
for x264, try --x264-speedcontrol, which will try to adjust the preset
dynamically for maximum quality, at the expense of somewhat higher delay.

See --help for more information on options in general.

The name “Nageru” is a play on the Japanese verb 投げる (nageru), which means
to throw or cast. (I also later learned that it could mean to face defeat or
give up, but that's not the intended meaning.)


Nageru's home page is at https://nageru.sesse.net/, where you can also find
contact information and link to the latest version.


Legalese: TL;DR: Everything is GPLv3-or-newer compatible, and see
Intel's copyright license at quicksync_encoder.h.


Nageru is Copyright (C) 2015 Steinar H. Gunderson <steinar+nageru@gunderson.no>.
Portions Copyright (C) 2003 Rune Holm.
Portions Copyright (C) 2010-2015 Fons Adriaensen <fons@linuxaudio.org>.
Portions Copyright (C) 2012-2015 Fons Adriaensen <fons@linuxaudio.org>.
Portions Copyright (C) 2008-2015 Fons Adriaensen <fons@linuxaudio.org>.
Portions Copyright (c) 2007-2013 Intel Corporation. All Rights Reserved.


This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


Portions of quicksync_encoder.h and quicksync_encoder.cpp:

Copyright (c) 2007-2013 Intel Corporation. All Rights Reserved.

Permission is hereby granted, free of charge, to any person obtaining a
copy of this software and associated documentation files (the
"Software"), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sub license, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice (including the
next paragraph) shall be included in all copies or substantial portions
of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT.
IN NO EVENT SHALL PRECISION INSIGHT AND/OR ITS SUPPLIERS BE LIABLE FOR
ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


All files in decklink/:

Copyright (c) 2009 Blackmagic Design
Copyright (c) 2015 Blackmagic Design

Permission is hereby granted, free of charge, to any person or organization
obtaining a copy of the software and accompanying documentation covered by
this license (the "Software") to use, reproduce, display, distribute,
execute, and transmit the Software, and to prepare derivative works of the
Software, and to permit third-parties to whom the Software is furnished to
do so, all subject to the following:

The copyright notices in the Software and this entire statement, including
the above license grant, this restriction and the following disclaimer,
must be included in all copies of the Software, in whole or in part, and
all derivative works of the Software, unless such copies or derivative
works are solely in the form of machine-executable object code generated by
a source language processor.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE, TITLE AND NON-INFRINGEMENT. IN NO EVENT
SHALL THE COPYRIGHT HOLDERS OR ANYONE DISTRIBUTING THE SOFTWARE BE LIABLE
FOR ANY DAMAGES OR OTHER LIABILITY, WHETHER IN CONTRACT, TORT OR OTHERWISE,
ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
DEALINGS IN THE SOFTWARE.


NEWS
Nageru 1.6.2, July 16th, 2017

  - Various smaller Kaeru fixes, mostly around metrics. Also,
    you can now adjust the x264 bitrate in Kaeru (in 100 kbit/sec
    increments) by sending SIGUSR1 (higher) or SIGUSR2 (lower).


Nageru 1.6.1, July 9th, 2017

  - Add native export of Prometheus metrics.

  - Rework the frame queue drop algorithm. The new one should handle tricky
    situations much better, especially when a card is drifting very slowly
    against the master timer.

  - Add Kaeru, an experimental transcoding tool based on Nageru code.
    Kaeru can run headless on a server without a GPU to transcode a
    Nageru stream into a lower-bitrate one, replacing VLC.

  - Work around a bug in some versions of NVIDIA's OpenGL drivers that would
    crash Nageru after about three hours (fix in cooperation with Movit).

  - Fix a crash with i965-va-driver 1.8.x.

  - Reduce mutex contention in certain critical places, causing lower tail
    latency in the mixer.


Nageru 1.6.0, May 29th, 2017

  - Add support for having videos (from file or from URL) as a separate
    input channels, albeit with some limitations. Apart from the obvious use of
    looping pause clips or similar, this can be used to integrate with CasparCG;
    see the manual for more details.

  - Add a frame analyzer (accessible from the Video menu) containing an
    RGB histogram and a color dropped tool. This is useful in calibrating
    video chains by playing back a known signal. Note that this adds a
    dependency on QCustomPlot.

  - Allow overriding Y'CbCr input interpretation, for inputs that don't
    use the correct settings. Also, Rec. 601 is now used by default instead
    of Rec. 709 for SD resolutions.

  - Support other sample rates than 48000 Hz from bmusb.


Nageru 1.5.0, April 5th, 2017

  - Support for low-latency HDMI/SDI output in addition to (or instead of) the
    stream. This currently only works with DeckLink cards, not bmusb. See the
    manual for more information.

  - Support changing the resolution from the command line, instead of locking
    everything to 1280x720.

  - The A/V sync code has been rewritten to be more in line with Fons
    Adriaensen's original paper. It handles several cases much better,
    in particular when trying to match 59.94 and 60 Hz sources to each other.
    However, it might occasionally need a few extra seconds on startup to
    lock properly if startup is slow.

  - Add support for using x264 for the disk recording. This makes it possible,
    among other things, to run Nageru on a machine entirely without VA-API
    support.

  - Support for 10-bit Y'CbCr, both on input and output. (Output requires
    x264 disk recording, as Quick Sync Video does not support 10-bit H.264.)
    This requires compute shader support, and is in general a little bit
    slower on input and output, due to the extra amount of data being shuffled
    around. Intermediate precision is 16-bit floating-point or better,
    as before.

  - Enable input mode autodetection for DeckLink cards that support it.
    (bmusb mode has always been autodetected.)

  - Add functionality to add a time code to the stream; useful for debugging
    latency.

  - The live display is now both more performant and of higher image quality.

  - Fix a long-standing issue where the preview displays would be too bright
    when using an NVIDIA GPU. (This did not affect the finished stream.)

  - Many other bugfixes and small improvements.


Nageru 1.4.2, November 24th, 2016

  - Fix a thread race that would sometimes cause x264 streaming to go awry.


Nageru 1.4.1, November 6th, 2016

  - Various bugfixes.


Nageru 1.4.0, October 26th, 2016

  - Support for multichannel (or more accurately, multi-bus) audio,
    choosable from the UI or using the --multichannel command-line
    flag. In multichannel mode, you can take in inputs from multiple
    different sources (or different channels on the same source, for
    multichannel sound cards), apply effects to them separately and then
    mix them together. This includes both audio from the video cards
    as well as ALSA inputs, including hotplug. Ola Gundelsby contributed
    invaluable feedback on this feature throughout the entire
    development cycle.

  - Support for having MIDI controllers control various aspects of the
    audio UI, with relatively flexible mapping. Note that different
    MIDI controllers can vary significantly in what protocol they speak,
    so Nageru will not necessarily work with all. (The primary testing
    controller has been the Akai MIDImix, and a pre-made mapping for
    that is included. The Korg nanoKONTROL2 has also been tested and
    works, but it requires some Korg-specific SysEx commands to make
    the buttons and lights work.)

  - Add a disk space indicator to the main window.

  - Various bugfixes. In particular, an issue where the audio would pitch
    up sharply after a series of many dropped frames has been fixed.


Nageru 1.3.4, August 2nd, 2016

  - Various bugfixes.


Nageru 1.3.3, July 27th, 2016

  - Various changes to make distribution packaging easier; in particular,
    theme data can be picked up from /usr/local/share/nageru.

  - Fix various FFmpeg deprecation warnings, now that we need FFmpeg
    3.1 for other reasons anyway.


Nageru 1.3.2, July 23rd, 2016

  - Allow limited hotplugging (unplugging and replugging) of USB cards.
    You can use the new command-line option --num-fake-cards (-C) to add
    fake cards that show only a single color and that will be replaced
    by real cards as you plug them in; you can also unplug cards and have
    them be replaced by fake cards. Fake cards can also be used for testing
    Nageru without actually having any video cards available.

  - Add Metacube timestamping of every keyframe, for easier detection of
    streams not keeping up. Works with the new timestamp feature of
    Cubemap 1.3.1. Will be ignored (save for some logging) in older

defs.h
#ifndef _DEFS_H
#define _DEFS_H

#define OUTPUT_FREQUENCY 48000  // Currently needs to be exactly 48000, since bmusb outputs in that.
#define MAX_FPS 60
#define FAKE_FPS 25  // Must be an integer.
#define MAX_VIDEO_CARDS 16
#define MAX_ALSA_CARDS 16
#define MAX_BUSES 256  // Audio buses.

// For deinterlacing. See also comments on InputState.
#define FRAME_HISTORY_LENGTH 5

#define AUDIO_OUTPUT_CODEC_NAME "pcm_s32le"
#define DEFAULT_AUDIO_OUTPUT_BIT_RATE 0
#define DEFAULT_X264_OUTPUT_BIT_RATE 4500  // 5 Mbit after making room for some audio and TCP overhead.

#define LOCAL_DUMP_PREFIX "record-"
#define LOCAL_DUMP_SUFFIX ".nut"
#define DEFAULT_STREAM_MUX_NAME "nut"  // Only for HTTP. Local dump guesses from LOCAL_DUMP_SUFFIX.
#define MUX_OPTS { \
	/* Make seekable .mov files. */ \
	{ "movflags", "empty_moov+frag_keyframe+default_base_moof" }, \
	\
	/* Make for somewhat less bursty stream output when using .mov. */ \
	{ "frag_duration", "125000" }, \
	\
	/* Keep nut muxer from using unlimited amounts of memory. */ \
	{ "write_index", "0" } \
}

// In bytes. Beware, if too small, stream clients will start dropping data.
// For mov, you want this at 10MB or so (for the reason mentioned above),
// but for nut, there's no flushing, so such a large mux buffer would cause
// the output to be very uneven.
#define MUX_BUFFER_SIZE 10485760

// In number of frames. Comes in addition to any internal queues in x264
// (frame threading, lookahead, etc.).
#define X264_QUEUE_LENGTH 50

#define X264_DEFAULT_PRESET "ultrafast"
#define X264_DEFAULT_TUNE "film"

#endif  // !defined(_DEFS_H)
